File: benchmarks/04-solve/solve.m
(* ==========================================================================
   Experiment 04 -- Solve: polynomial, systems, transcendental
   ==========================================================================
   WHAT IT MEASURES.  src/solve.c and src/poly/solvepoly.c -- root isolation for
   univariate polynomials, elimination for systems, and the inverse-function
   branch for transcendental equations.

   VALUE CHECKS ARE COUNTS AND SUMS, not solution sets.  Two systems legitimately
   order roots differently and spell radicals differently, so the check is the
   NUMBER of solutions plus the rounded SUM of the roots -- both invariant under
   ordering and under algebraic form.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"Solve", "NSolve", "Reduce", "Eliminate", "Roots"}];

(* 1. A factorable sextic: rational roots, no radicals needed. *)
e1 = Product[x - i, {i, 1, 6}] == 0;
bench["Solve factorable sextic", Solve[e1, x];];
check["Solve factorable sextic", Length[Solve[e1, x]]];

(* 2. A quartic needing radicals. *)
e2 = x^4 - 10 x^2 + 1 == 0;
bench["Solve quartic with radicals", Solve[e2, x];];
check["Solve quartic with radicals", Length[Solve[e2, x]]];

(* 3. A 2x2 polynomial system: elimination. *)
bench["Solve 2x2 polynomial system",
  Solve[{x^2 + y^2 == 25, x - y == 1}, {x, y}];];
check["Solve 2x2 polynomial system",
  Length[Solve[{x^2 + y^2 == 25, x - y == 1}, {x, y}]]];

(* 4. A 3x3 linear system: the cheap baseline.  If this row is slow the cost is
   in expression handling, not in any solving algorithm. *)
bench["Solve 3x3 linear system",
  Solve[{x + y + z == 6, x - y == 1, y - z == 1}, {x, y, z}];];
check["Solve 3x3 linear system",
  Length[Solve[{x + y + z == 6, x - y == 1, y - z == 1}, {x, y, z}]]];

(* 5. Numeric roots of a degree-40 polynomial. *)
p40 = denseUPoly[40, x];
bench["NSolve degree 40", NSolve[p40 == 0, x];];
check["NSolve degree 40", Length[NSolve[p40 == 0, x]]];

(* 6. Sum of the roots of the sextic -- pins the VALUES, not just the count. *)
bench["Solve sextic, root sum", Total[x /. Solve[e1, x]];];
check["Solve sextic, root sum", Round[Total[x /. Solve[e1, x]]]];

File: benchmarks/04-solve/solve.py
#!/usr/bin/env python3
"""Experiment 04 -- Solve: polynomial, systems, transcendental (sympy column).

Same six kernels as ``solve.m``, same order.

VALUE CHECKS ARE COUNTS AND SUMS, not solution sets: two systems legitimately
order roots differently and spell radicals differently, so the check is the
number of solutions plus the rounded sum of the roots -- both invariant under
ordering and under algebraic form.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import sympy

from harness import bench, check, require
from data import dense_upoly

require(["sympy", "sympy:solve", "sympy:nroots", "sympy:solveset"])

x, y, z = sympy.symbols("x y z")

e1 = sympy.prod([x - i for i in range(1, 7)])
bench("Solve factorable sextic", lambda: sympy.solve(sympy.Eq(e1, 0), x))
check("Solve factorable sextic", len(sympy.solve(sympy.Eq(e1, 0), x)))

e2 = x ** 4 - 10 * x ** 2 + 1
bench("Solve quartic with radicals", lambda: sympy.solve(sympy.Eq(e2, 0), x))
check("Solve quartic with radicals", len(sympy.solve(sympy.Eq(e2, 0), x)))

sys2 = [sympy.Eq(x ** 2 + y ** 2, 25), sympy.Eq(x - y, 1)]
bench("Solve 2x2 polynomial system", lambda: sympy.solve(sys2, [x, y]))
check("Solve 2x2 polynomial system", len(sympy.solve(sys2, [x, y])))

sys3 = [sympy.Eq(x + y + z, 6), sympy.Eq(x - y, 1), sympy.Eq(y - z, 1)]
# sympy returns a single dict for a determined linear system; Wolfram returns a
# list of one rule-list. Both have "one solution", so the check is 1 either way.
bench("Solve 3x3 linear system", lambda: sympy.solve(sys3, [x, y, z]))
_s3 = sympy.solve(sys3, [x, y, z])
check("Solve 3x3 linear system", 1 if isinstance(_s3, dict) else len(_s3))

p40 = sympy.Poly(dense_upoly(40, x), x)
bench("NSolve degree 40", lambda: p40.nroots(n=15, maxsteps=200))
check("NSolve degree 40", len(p40.nroots(n=15, maxsteps=200)))

bench("Solve sextic, root sum", lambda: sum(sympy.solve(sympy.Eq(e1, 0), x)))
check("Solve sextic, root sum", int(sum(sympy.solve(sympy.Eq(e1, 0), x))))
